internal/function/diff.go

```diff
@@ -0,0 +1,292 @@
+	}
+
+	return d, nil
+}
+
+func (d *Diff) String() string {
+	if d.from == nil {
+		return fmt.Sprintf("diff(%s, %s)", d.repo, d.to)
+	}
+
+	return fmt.Sprintf("diff(%s, %s, %s)", d.repo, d.from, d.to)
+}
+
+// Type implements the sql.Expression interface.
+func (*Diff) Type() sql.Type {
+	return sql.Array(sql.JSON)
+}
+
+// TransformUp implements the Expression interface.
+func (d *Diff) TransformUp(fn sql.TransformExprFunc) (sql.Expression, error) {
+	repo, err := d.repo.TransformUp(fn)
+	if err != nil {
+		return nil, err
+	}
+
+	to, err := d.to.TransformUp(fn)
+	if err != nil {
+		return nil, err
+	}
+
+	if d.from == nil {
+		return fn(&Diff{repo: repo, to: to})
+	}
+
+	from, err := d.from.TransformUp(fn)
+	if err != nil {
+		return nil, err
+	}
+	return fn(&Diff{repo: repo, from: from, to: to})
+}
+
+// Children implements the Expression interface.
+func (d *Diff) Children() []sql.Expression {
+	if d.from == nil {
+		return []sql.Expression{d.repo, d.to}
+	}
+	return []sql.Expression{d.repo, d.from, d.to}
+}
+
+// IsNullable implements the Expression interface.
+func (*Diff) IsNullable() bool {
+	return false
+}
+
+// Resolved implements the Expression interface.
+func (d *Diff) Resolved() bool {
+	return d.to.Resolved() && d.to.Resolved() && (d.from == nil || d.from.Resolved())
+}
+
+// Eval implements the sql.Expression interface.
+func (d *Diff) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
+	span, ctx := ctx.Span("gitbase.Diff")
+	defer span.Finish()
+
+	repo, err := d.getRepo(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+
+	from, to, err := d.getCommits(ctx, repo, row)
+
+	return diff(ctx, from, to)
+}
+
+func (d *Diff) getRepo(ctx *sql.Context, r sql.Row) (*gitbase.Repository, error) {
+	repoID, err := exprToString(ctx, d.repo, r)
+	if err != nil {
+		return nil, err
+	}
+	s, ok := ctx.Session.(*gitbase.Session)
+	if !ok {
+		return nil, gitbase.ErrInvalidGitbaseSession.New(ctx.Session)
+	}
+	return s.Pool.GetRepo(repoID)
+}
+
+func (d *Diff) getCommits(ctx *sql.Context, repo *gitbase.Repository, row sql.Row) (*object.Commit, *object.Commit, error) {
+	commitHash, err := exprToString(ctx, d.to, row)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	to, err := repo.CommitObject(plumbing.NewHash(commitHash))
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var from *object.Commit
+	if d.from != nil {
+		commitHash, err = exprToString(ctx, d.from, row)
+		if err != nil {
+			return nil, nil, err
+		}
+		from, err = repo.CommitObject(plumbing.NewHash(commitHash))
+	} else {
+		from, err = to.Parent(0)
+	}
+	if err != nil && err != object.ErrParentNotFound {
+		return nil, nil, err
+	}
+
+	return from, to, nil
+}
+
+// diff compares two commits and returns a slice of diff items.
+// It compares changes (file by file) in parallel.
+// Another go routine collects diff items.
+// The function ignores binary files and equal operations.
+func diff(ctx context.Context, from, to *object.Commit) ([]*diffItem, error) {
```

Comment:
Instead of writing this from scratch we should use [Commit.PatchContext](https://godoc.org/gopkg.in/src-d/go-git.v4/plumbing/object#Commit.PatchContext), in this case, you will find several problems to be fixed on `go-git` such as diff on binaries, or some other bugs, but this really worth it, since we are going to have many `diff` functions on gitbase

Reply:
@mcuadros - Ok, I've switched to `PatchContext` but I did it on Tree level, because on `object.Commit` level there is no check if `from` and/or `to` commits are `nil`, and `from` can be `nil` if there is no parent (the first commit).